portfolio/src/main/webapp/script.js

```diff
@@ -28,103 +28,233 @@ function addRandomGreeting() {
+            }
+        }
 
-    while (source.charAt(index) != '<'){
-        index--;
     }
 
-    var startIndex = index;
+    return matches;
+}
+
+/**
+ * Clears the current highlighted text nodes
+ * Will optimize to actually remove the span rather than the class
+ */
+function clear(){
+    let highlightedList = document.querySelectorAll('.highlight');
 
-    return source.substring(startIndex, endIndex)
+    highlightedList.forEach(function(el) {
+        el.replaceWith(el.innerText);
+    });
 }
 
 /**
-  * Gets HTML id from given word
-  */
-function retrieveId(string, source){
-    var searchForId = string.search("id=\"");
-    if (!searchForId){
+ * Handles clearing old nodes and end of search run,
+ * as well as calls the highlighting and scrolling fns
+ * @param {object} matches - Array of matching nodes in DOM tree.
+ * @param {string} input - Input retrieved from search.
+ * @param {number} index - Points to the node in the array we want to highlight.
+ */
+function highlightHandler(matches, input, index){
+    if (index === matches.length && index != 0){
+        clear();
+        setHidden('nextbutton', true);
+        match_index = 0; // resetting the index
+        document.body.scrollIntoView({behavior: "smooth"}); // back to the top
         return;
     }
-    const startIndex = searchForId+4;
-    var retString = string.charAt(startIndex);
 
-    for(var i = startIndex+1; string.charAt(i) != "\""; i++){
-        retString += string.charAt(i);
+    if (index > 0){
+        clear();
     }
+    
+    // highlight and scroll to every element
+    highlighter(matches[index], input);
+    
+    let els = document.getElementsByClassName('highlight');
+    els[0].scrollIntoView({behavior: "smooth"}); 
 
-    return retString;
+    match_index++; // go onto the next value
 }
 
+
 /**
-  * Runs the search by getting the source HTML, getting the appropriate tag, and scrolling towards that id
-  */
-function runSearch(input){
-    var source = retrieveHtml();
+ * Highlights a given text node
+ * @param {object} node - The node in the tree I'm highlighting.
+ * @param {string} input - The string from the searchbar, used to find matches in node.
+ */
+function highlighter(node, input){
+    const nodeText = node.textContent;
 
-    var foundIndex = source.indexOf(input);
-    if (foundIndex == -1){return};
+    let foundIndices = findSubstringIndices(input, nodeText);
+
+    if (foundIndices == [] || foundIndices == null){
+        return;
+    }
+
+    let highlightedString = node.textContent.substring(foundIndices[0], input.length+foundIndices[0]);
+    let fragments = nodeText.split(highlightedString);
+
+    const newNode = document.createElement('span');
+    newNode.setAttribute('class', 'highlight');
+    newNode.textContent = highlightedString;
+
+    let rejoinedNode = [];
+    for (let i = 0; i < fragments.length; i++){
+        rejoinedNode.push(fragments[i]);
+        if (i != fragments.length-1){
```

Comment:
Just iterate up to `fragments.length-1` instead of checking this if every time:
```
// Zips text fragments with highlighted spans
let zippedNodes [];
for (let i = 0; i < fragments.length - 1; i++) {
  zippedNodes.push(fragments[i]);
  zippedNodes.push(highlightedNodes[i]);
}
zippedNodes.push(fragments[fragments.length - 1]);
```